cub/cub/device/dispatch/dispatch_transform.cuh

```diff
@@ -0,0 +1,995 @@
+    {
+      tile_size = len - global_offset;
+    }
+    int smem_offset            = 0;
+    std::uint32_t total_copied = 0;
+
+#  ifdef __cpp_fold_expressions // C++17
+    // Order of evaluation is always left-to-right here. So smem_offset is updated in the right order.
+    (..., copy_ptr_set(bar, tile_size, tile_stride, smem, smem_offset, total_copied, global_offset, pointers));
+#  else
+    // Order of evaluation is also left-to-right
+    int dummy[] = {
+      (copy_ptr_set(bar, tile_size, tile_stride, smem, smem_offset, total_copied, global_offset, pointers), 0)..., 0};
+    (void) dummy;
+#  endif
+
+    ::cuda::ptx::mbarrier_arrive_expect_tx(
+      ::cuda::ptx::sem_release, ::cuda::ptx::scope_cta, ::cuda::ptx::space_shared, &bar, total_copied);
+  }
+  __syncthreads();
+
+  while (!::cuda::ptx::mbarrier_try_wait_parity(&bar, 0))
+  {
+  }
+  // Intentionally use unroll 1. This tends to improve performance.
+#  pragma unroll 1
+  for (int j = 0; j < num_elem_per_thread; ++j)
+  {
+    // ahendriksen: We do not define g_idx in terms of smem_idx. Doing this results in sub-optimal codegen. The common
+    // sub-expression elimination logic is smart enough to remove the redundant computations.
+    const int smem_idx = j * blockDim.x + threadIdx.x;
+    const Offset g_idx = global_offset + j * blockDim.x + threadIdx.x;
+    if (g_idx < len)
+    {
+      int smem_offset = 0;
+      out[g_idx]      = f(fetch_operand(tile_stride, smem, smem_offset, smem_idx, pointers)...);
+    }
+  }
+#endif // CUB_PTX_ARCH >= 900
+}
+
+// Type erasing "union" for kernel arguments. We cannot use a real union, because it would not be trivially copyable in
+// case It is not trivially copyable (e.g. thrust::constant_iterator<custom_numeric>)
+template <typename It>
+struct kernel_arg
+{
+  using PS                               = ptr_set<const value_t<It>>;
+  static constexpr std::size_t alignment = ::cuda::std::max(alignof(It), alignof(PS)); // need extra variable for GCC<9
+  alignas(alignment) char storage[::cuda::std::max(sizeof(It), sizeof(PS))];
+
+  template <typename T>
+  _CCCL_HOST_DEVICE T& aliased_storage()
+  {
+    return *reinterpret_cast<T*>(storage);
+  }
+};
+
+template <typename It>
+_CCCL_HOST_DEVICE auto make_iterator_kernel_arg(It it) -> kernel_arg<It>
+{
+  kernel_arg<It> arg;
+  cub::detail::uninitialized_copy_single(&arg.template aliased_storage<It>(), it);
+  return arg;
+}
+
+template <typename It>
+_CCCL_HOST_DEVICE auto make_ptr_set_kernel_arg(It ptr) -> kernel_arg<It>
+{
+  kernel_arg<It> arg;
+  using T = value_t<It>;
+  cub::detail::uninitialized_copy_single(
+    &arg.template aliased_storage<ptr_set<const T>>(),
+    ptr_set<const T>{
+      round_down_ptr_128(ptr),
+      round_up_16(offset_to_aligned_ptr_128(ptr) * sizeof(T)),
+      // (ptr - ptr_base) * sizeof(T) rounded up to nearest multiple of 16
+      static_cast<uint32_t>(sizeof(T) * (ptr - round_down_ptr_128(ptr))),
+    });
```

Comment:
question: should this be `make_ublkcp_ptr_set`?

Reply:
This was copied from `make_ublkcp_ptr_set` and I forgot to delete it.